locking/cache.go

```diff
@@ -0,0 +1,213 @@
+package locking
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/github/git-lfs/api"
+
+	"github.com/boltdb/bolt"
+	"github.com/github/git-lfs/config"
+)
+
+var (
+	dbInit             sync.Once
+	lockDb             *bolt.DB
+	pathToIdBucketName = []byte("pathToId")
+	idToPathBucketName = []byte("idToPath")
+)
+
+func initDb() error {
+	// Open on demand - bolt will lock this file & other processes trying to access it
+	// we'll wait a max 5 seconds
+	// TODO: could have option to open read-only to take shared lock
+	var initerr error
+	dbInit.Do(func() {
+		lockDir := filepath.Join(config.LocalGitStorageDir, "lfs")
+		os.MkdirAll(lockDir, 0755)
+		lockFile := filepath.Join(lockDir, "db.lock")
+		lockDb, initerr = bolt.Open(lockFile, 0644, &bolt.Options{Timeout: 5 * time.Second})
+	})
+	if initerr != nil {
+		// TODO maybe suggest re-initialising lock cache
+		return initerr
+	}
+
+	// Very important that Cleanup() is called before shutdown
+
+	return nil
+}
+
+func Cleanup() error {
+	if lockDb != nil {
+		return lockDb.Close()
+	}
+	return nil
+}
+
+// Run a read-only lock database function
+// Deals with initialisation and function is a transaction
+// Can run in a goroutine if needed
+func runLockDbReadOnlyFunc(f func(tx *bolt.Tx) error) error {
+
+	if err := initDb(); err != nil {
+		return err
+	}
+
+	return lockDb.View(f)
+}
+
+// Run a read-write lock database function
+// Deals with initialisation and function is a transaction
+// Can run in a goroutine if needed
+func runLockDbFunc(f func(tx *bolt.Tx) error) error {
+
+	if err := initDb(); err != nil {
+		return err
+	}
+
+	// Use Batch() to improve write performance and goroutine friendly
+	return lockDb.Batch(f)
+}
+
+// This file caches active locks locally so that we can more easily retrieve
+// a list of locally locked files without consulting the server
+// This only includes locks which the local committer has taken, not all locks
+
+// Cache a successful lock for faster local lookup later
+func cacheLock(filePath, id string) error {
+	return runLockDbFunc(func(tx *bolt.Tx) error {
+		path2id, err := tx.CreateBucketIfNotExists(pathToIdBucketName)
+		if err != nil {
+			return err
+		}
+		id2path, err := tx.CreateBucketIfNotExists(idToPathBucketName)
+		if err != nil {
+			return err
+		}
```

Comment:
>  didn't use api.Lock because I prefer to avoid introducing transitive dependencies in function signatures

:+1: I'm thinking we could even eventually get rid of `api`. Callers should just be using packages like `locking`, `transfer`, `batch`, etc.